### PR TITLE
cephfs-journal-tool: Adding [--repair] option for cephfs-journal-tool make it can recover all journal as much as possible.

### PR DESCRIPTION
**Pull request by yunfeiguan**
If entry size is corrupt may lost more than one journal event.

Fixes: http://tracker.ceph.com/issues/21888
Signed-off-by: Guan yunfei <yunfei.guan@xtaotech.com>

### CODE REVIEW

**Comment by batrick** [comment]
@yunfeiguan , please update the commit header to prefix with "cephfs-journal-tool: ". Guidelines here: https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#title-of-pull-requests-and-title-of-commits

**Comment by yunfeiguan** [comment]
@jcsp help to review please, thank you.

**Comment by ukernel** [comment]
If entry size is corrupted, I think there is no point to assume the rest part of the entry is valid.  I think it's better just push bytes back onto start of read_buf and just advance one byte to read_offset.

**Comment by batrick** [comment]
@yunfeiguan please address @ukernel 's comments.

**Comment by batrick** [comment]
Closing this due to inactivity.

**Comment by yunfeiguan** [comment]
@batrick sorry, there is no time to pay attention to this PR some days ago.

**Comment by yunfeiguan** [comment]
@ukernel There is really not much opportunity to save the entry,but it will leads a worse case which will make read_offset move to a unpredictable position.Thus we may miss a lot of entry in next read.

**Comment by batrick** [comment]
@yunfeiguan if you would like to revisit this, please go ahead and rebase.

**Comment by yunfeiguan** [comment]
@batrick rebased done, help to review,tthanks.

File: src/tools/cephfs/JournalTool.cc [hunk 1]
```diff
@@ -367,6 +369,15 @@ int JournalTool::main_event(std::vector<const char*> &argv)
     return r;
   }
 
+  // Check if need to repair
+  for (std::vector<const char*>::iterator iter = argv.begin(); iter != argv.end(); iter++) {
```
Comment (batrick) [inline src/tools/cephfs/JournalTool.cc]: Please use:

`for (const auto& s : argv) {`

File: src/tools/cephfs/JournalScanner.cc [hunk 2]
```diff
@@ -252,7 +368,23 @@ int JournalScanner::scan_events()
 
         if (!readable) {
           // Out of data, continue to read next object
-          break;
+	  if (repair_corrupted_entry) {
```
Comment (batrick) [inline src/tools/cephfs/JournalScanner.cc]: indentation is inconsistent. Please only use spaces.

File: src/tools/cephfs/JournalTool.cc (hunk 1 above)
Comment (batrick) [reply src/tools/cephfs/JournalTool.cc]: Also, why is this option code not in `JournalTool::main`?

File: src/tools/cephfs/JournalScanner.cc [hunk 3]
```diff
@@ -23,6 +23,121 @@
...
+{
+  if (buf.length() == 0)
+    return false;
+ 
+  bufferlist temp_buf;
+  buf.copy(0, buf.length(), temp_buf);
+  do {
+    static int sentinel_num = 0;
+    bufferlist::iterator p = temp_buf.begin();
+    static int _off = 0;
```
Comment (batrick) [inline src/tools/cephfs/JournalScanner.cc]: Please rework the code so these static variables are not necessary. Either move the variables to `JournalScanner` or pass the variables in as arguments.

File: src/tools/cephfs/JournalTool.cc (hunk 1 above)
Comment (yunfeiguan) [reply src/tools/cephfs/JournalTool.cc]: the option only be used in JournalTool::main_event, so i think we don't need put it in JournalTool::main.Secondly, we need pass the repaire option to JournalTool::main_event or define a member variable of JournalTool if we code this option in JournalTool::main.

File: src/tools/cephfs/JournalScanner.h [hunk 4]
```diff
@@ -80,11 +86,16 @@ class JournalScanner
...
+  bool buf_is_available(bufferlist& buf);
+  void repair_journal(bufferlist& buf, uint64_t offset);
   void report(std::ostream &out) const;
 
   std::string obj_name(uint64_t offset) const;
   std::string obj_name(inodeno_t ino, uint64_t offset) const;
 
+  bool repair_corrupted_entry;
+  int sentinel_num;
+  int _off;
```
Comment (ukernel) [inline src/tools/cephfs/JournalScanner.h]: why define seninel_num and _off as class menbers

**Comment by ukernel** [comment]
why I mean is if entry size or start_ptr is incorrect, we just ignore current entry. Search for the next sentinel instead.

**Comment by yunfeiguan** [comment]
@ukernel @batrick help review please! thank you!

**Pull request closed without merging**